zerver/tests/test_message_fetch.py

```diff
@@ -2675,6 +2675,160 @@ def _update_tsvector_index(self) -> None:
+        self.setup_web_public_test(5)
+        channel = get_stream("web-public-channel", realm)
+        assert channel.recipient_id is not None
+        message_ids = messages_for_topic(realm.id, channel.recipient_id, "test").values_list(
+            "id", flat=True
+        )
+
+        web_public_narrow = [
+            dict(operator="channels", operand="web-public", negated=False),
+            dict(operator="channel", operand="web-public-channel"),
+            dict(operator="topic", operand="test"),
+            dict(operator="with", operand=message_ids[0]),
+        ]
+        post_params = {
+            "anchor": 0,
+            "num_before": 0,
+            "num_after": 5,
+            "narrow": orjson.dumps(web_public_narrow).decode(),
+        }
+
+        result = self.client_get("/json/messages", dict(post_params))
+        self.verify_web_public_query_result_success(result, 5)
+
+        # Test `with` operator ineffective when spectator does not have
+        # access to message, by trying to access the same set of messages
+        # but when the spectator access is not allowed.
+        do_set_realm_property(hamlet.realm, "enable_spectator_access", False, acting_user=hamlet)
+
+        result = self.client_get("/json/messages", dict(post_params))
+        self.check_unauthenticated_response(result)
+
+        # Test request with multiple `with` operators raises
+        # InvalidOperatorCombinationError
+        self.login("iago")
+        iago = self.example_user("iago")
+        msg_id_1 = self.send_stream_message(iago, "Verona")
+        msg_id_2 = self.send_stream_message(iago, "Scotland")
+
+        narrow = [
+            dict(operator="channel", operand="Verona"),
+            dict(operator="with", operand=msg_id_1),
+            dict(operator="topic", operand="test"),
+            dict(operator="with", operand=msg_id_2),
+        ]
+        post_params = {
+            "anchor": msg_id_1,
+            "num_before": 0,
+            "num_after": 5,
+            "narrow": orjson.dumps(narrow).decode(),
+        }
+        result = self.client_get("/json/messages", dict(post_params))
+        self.assert_json_error(result, "Invalid narrow operator: Duplicate 'with' operators.")
+
+        # Test request with an invalid message id for `with` operator
+        # raises InvalidOperatorCombinationError
+        msg_id = self.send_stream_message(iago, "Verona", topic_name="Invalid id")
+        narrow = [
+            dict(operator="channel", operand="Verona"),
+            dict(operator="topic", operand="Invalid id"),
+            dict(operator="with", operand="3.2"),
+        ]
+        post_params = {
+            "anchor": msg_id,
+            "num_before": 0,
+            "num_after": 5,
+            "narrow": orjson.dumps(narrow).decode(),
+        }
+        result = self.client_get("/json/messages", dict(post_params))
+        self.assert_json_error(result, "Invalid narrow operator: Invalid 'with' operator")
```

Comment:
Reading this test carefully, it really didn't test a lot of the behavior that's important to test. Went through and changed it to verify what we need to precisely.

``` diff
diff --git a/zerver/tests/test_message_fetch.py b/zerver/tests/test_message_fetch.py
index 6aee845db2..87ec7df219 100644
--- a/zerver/tests/test_message_fetch.py
+++ b/zerver/tests/test_message_fetch.py
@@ -2687,33 +2687,28 @@ class GetOldMessagesTest(ZulipTestCase):
 
         # Test `with` operator effective when targeting a topic with
         # message which can be accessed by the user.
-        msg_id = self.send_stream_message(iago, "dev team")
+        msg_id = self.send_stream_message(iago, "dev team", topic_name="test")
 
         narrow = [
             dict(operator="channel", operand="dev team"),
-            dict(operator="topic", operand="test"),
+            dict(operator="topic", operand="other_topic"),
             dict(operator="with", operand=msg_id),
         ]
         results = self.get_and_check_messages(dict(narrow=orjson.dumps(narrow).decode()))
         self.assertEqual(results["messages"][0]["id"], msg_id)
+        # Notably we returned the message with its actual topic.
+        self.assertEqual(results["messages"][0]["subject"], "test")
 
-        # Test `with` operator returns messages with corrected narrow,
-        # containing message with id equivalent to `with` operand.
-        msg_ids = [
-            self.send_stream_message(iago, "dev team", topic_name="test 1") for _ in range(2)
-        ]
+        # Test `with` operator without channel/topic operators.
         narrow = [
-            dict(operator="channel", operand="Verona"),
-            dict(operator="topic", operand="test 1"),
-            dict(operator="with", operand=msg_ids[0]),
+            dict(operator="with", operand=msg_id),
         ]
         results = self.get_and_check_messages(dict(narrow=orjson.dumps(narrow).decode()))
-        for msg in results["messages"]:
-            self.assertIn(msg["id"], msg_ids)
+        self.assertEqual(results["messages"][0]["id"], msg_id)
 
         # Test `with` operator ineffective when targeting a topic with
         # message that can not be accessed by the user.
-        msg_id = self.send_stream_message(iago, "dev team")
+        # Since !history_public_to_subscribers, hamlet cannot view.
         self.subscribe(hamlet, "dev team")
         self.login("hamlet")
 
@@ -2725,17 +2720,46 @@ class GetOldMessagesTest(ZulipTestCase):
         results = self.get_and_check_messages(dict(narrow=orjson.dumps(narrow).decode()))
         self.assert_length(results["messages"], 0)
 
+        # Same result with topic specified incorrectly
+        narrow = [
+            dict(operator="channel", operand="dev team"),
+            dict(operator="topic", operand="wrong_guess"),
+            dict(operator="with", operand=msg_id),
+        ]
+        results = self.get_and_check_messages(dict(narrow=orjson.dumps(narrow).decode()))
+        self.assert_length(results["messages"], 0)
+
+        # If just with is specified, we get messages a la combined feed,
+        # but not the target message.
+        narrow = [
+            dict(operator="with", operand=msg_id),
+        ]
+        results = self.get_and_check_messages(dict(narrow=orjson.dumps(narrow).decode()))
+        self.assertNotIn(msg_id, [message["id"] for message in results["messages"]])
+
         # Test `with` operator is effective when targeting personal
         # messages with message id, and returns messages of that narrow.
-        self.login("iago")
+        #
+        # This will be relevant if we allow moving DMs in the future.
+        #
+        # First, attempt to view a message ID we can't access.
         msg_ids = [self.send_personal_message(iago, othello) for _ in range(2)]
-
         with_narrow = [
-            dict(operator="dm", operand=[othello.id]),
+            # Important: We pass the wrong conversation.
+            dict(operator="dm", operand=[hamlet.id]),
             dict(operator="with", operand=msg_ids[0]),
         ]
         results = self.get_and_check_messages(dict(narrow=orjson.dumps(with_narrow).decode()))
+        self.assertNotIn(msg_id, [message["id"] for message in results["messages"]])
 
+        # Now switch to a user how does have access.
+        self.login("iago")
+        with_narrow = [
+            # Important: We pass the wrong conversation.
+            dict(operator="dm", operand=[hamlet.id]),
+            dict(operator="with", operand=msg_ids[0]),
+        ]
+        results = self.get_and_check_messages(dict(narrow=orjson.dumps(with_narrow).decode()))
         for msg in results["messages"]:
             self.assertIn(msg["id"], msg_ids)
 
@@ -2748,7 +2772,8 @@ class GetOldMessagesTest(ZulipTestCase):
         msg_ids = [self.send_group_direct_message(iago, [cordelia, hamlet]) for _ in range(2)]
 
         with_narrow = [
-            dict(operator="dm", operand=[hamlet.id, cordelia.id]),
+            # Again, query the wrong conversation.
+            dict(operator="dm", operand=[hamlet.id]),
             dict(operator="with", operand=msg_ids[0]),
         ]
         results = self.get_and_check_messages(dict(narrow=orjson.dumps(with_narrow).decode()))
@@ -2769,7 +2794,8 @@ class GetOldMessagesTest(ZulipTestCase):
         web_public_narrow = [
             dict(operator="channels", operand="web-public", negated=False),
             dict(operator="channel", operand="web-public-channel"),
-            dict(operator="topic", operand="test"),
+            # Important: Pass a topic that does't contain the target message
+            dict(operator="topic", operand="wrong topic"),
             dict(operator="with", operand=message_ids[0]),
         ]
         post_params = {
@@ -2810,10 +2836,9 @@ class GetOldMessagesTest(ZulipTestCase):
             "narrow": orjson.dumps(narrow).decode(),
         }
         result = self.client_get("/json/messages", dict(post_params))
-        self.assert_json_error(result, "Invalid narrow operator: Duplicate 'with' operators.")
+        self.assert_json_error(result, "Invalid narrow operator combination: Duplicate 'with' operators.")
 
-        # Test request with an invalid message id for `with` operator
-        # raises InvalidOperatorCombinationError
+        # Test request with an invalid message id for `with` operator fails.
         msg_id = self.send_stream_message(iago, "Verona", topic_name="Invalid id")
         narrow = [
             dict(operator="channel", operand="Verona"),
```